File: Leetcode3070. Count submatrices with top-left element and sum less than k/Solution.cpp
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int n = grid.size(), m = grid[0].size();
        vector<int> cols(m);
        int res = 0;
        for (int i = 0; i < n; i++) {
            int rows = 0;
            for (int j = 0; j < m; j++) {
                cols[j] += grid[i][j];
                rows += cols[j];
                if (rows <= k) {
                    res++;
                }
            }
        }
        return res;
    }
};


// runtime - 0ms
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int n = grid.size(), m = grid[0].size();
        vector<int> cols(m);
        int res = 0;
        for (int i = 0; i < n; i++) {
            int rows = 0;
            for (int j = 0; j < m; j++) {
                cols[j] += grid[i][j];
                rows += cols[j];
                if (rows <= k) {
                    res++;
                }
            }
        }
        return res;
    }
};


// runtime - 1ms
/*class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        
    }
};*/
class Solution {
public:
    static int countSubmatrices(vector<vector<int>>& grid, int k) {
        const int r=grid.size(), c=grid[0].size();
        int cnt=0, brCol=c;
        if (grid[0][0]>k) return 0;// early stop
        cnt++;
        for(int j=1; j<c; j++){
            int& x=grid[0][j];
            x+=grid[0][j-1];
            if(x>k)// no need for computing for the rest cols
                { brCol=j; break;}
            cnt++;
        }
        for(int i=1; i<r; i++){
            grid[i][0]+=grid[i-1][0];
            if (grid[i][0]>k) break;
            cnt++;
            for(int j=1; j<brCol; j++){
                int& x=grid[i][j];
                x+=grid[i-1][j]+grid[i][j-1]-grid[i-1][j-1];
                if (x>k){
                    brCol=j; break;
                }
                cnt++;
            }
        }
        return cnt;
    }
};


// runtime - 2ms
class Solution {
public:
    static int countSubmatrices(vector<vector<int>>& grid, int k) {
        const int r=grid.size(), c=grid[0].size();
        int cnt=0, brCol=c;
        if (grid[0][0]>k) return 0;// early stop
        cnt++;
        for(int j=1; j<c; j++){
            int& x=grid[0][j];
            x+=grid[0][j-1];
            if(x>k)// no need for computing for the rest cols
                { brCol=j; break;}
            cnt++;
        }
        for(int i=1; i<r; i++){
            grid[i][0]+=grid[i-1][0];
            if (grid[i][0]>k) break;
            cnt++;
            for(int j=1; j<brCol; j++){
                int& x=grid[i][j];
                x+=grid[i-1][j]+grid[i][j-1]-grid[i-1][j-1];
                if (x>k){
                    brCol=j; break;
                }
                cnt++;
            }
        }
        return cnt;
    }
};


// runtime - 3ms
class Solution {
public:
    int countSubmatrices(const vector<vector<int>>& grid, const int k) {
        int ans = 0;
        vector<int> rsums(grid[0].size(), 0);

        for (size_t r = 0; r < grid.size(); ++r) {
            int rows = 0;

            for (size_t c = 0; c < grid[r].size(); ++c) {
                rsums[c] += grid[r][c];
                rows += rsums[c];
                //rsum += grid[r][c];
                if (rows <= k) {
                    //cout << r << "x" << c << " " << rsums[c] << endl;
                    ++ans;
                } else {
                    //sum -= k;
                    //break;
                }
            }
        }

        return ans;        
    }
};


// runtime - 4ms
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int n = grid.size(), m = grid[0].size();
        vector<int> cols(m);
        int res = 0;
        for (int i = 0; i < n; i++) {
            int rows = 0;
            for (int j = 0; j < m; j++) {
                cols[j] += grid[i][j];
                rows += cols[j];
                if (rows <= k) {
                    res++;
                }
            }
        }
        return res;
    }
};


// runtime - 5ms
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int n = grid.size(), m = grid[0].size();
        vector<int> cols(m);
        int res = 0;
        for (int i = 0; i < n; i++) {
            int rows = 0;
            for (int j = 0; j < m; j++) {
                cols[j] += grid[i][j];
                rows += cols[j];
                if (rows <= k) {
                    res++;
                }
            }
        }
        return res;
    }
};

// runtime - 6ms
class Solution {
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int ret = 0;
        int m = grid.size(), n = grid[0].size();
        {
            int tempSum = 0;
            int j=0;
            for (; j<n; ++j){
                grid[0][j] = (tempSum += grid[0][j]);
                if (grid[0][j]>k) break;
            }
            //cout << j << " ";
            ret += j;
        }
        for (int i=1; i<m; ++i){
            int tempSum = 0;
            int j=0;
            for (; j<n; ++j){
                grid[i][j] = (tempSum += grid[i][j]) + grid[i-1][j];
                if (grid[i][j]>k) break;
            }
            //cout << j << " ";
            ret += j;
        }
        return ret;
    }
};

File: Leetcode3070. Count submatrices with top-left element and sum less than k/Solution.py
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        n, m = len(grid), len(grid[0])
        cols = [0] * m
        res = 0

        for i in range(n):
            row_sum = 0
            for j in range(m):
                cols[j] += grid[i][j]
                row_sum += cols[j]
                if row_sum <= k:
                    res += 1

        return res


# runtime - 13ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        curWidth, curSum = len(grid[0]), sum(grid[0])
        while curSum > k:
            curWidth -= 1
            curSum -= grid[0][curWidth]
        result = curWidth
        for i in range(1, len(grid)):
            curSum += sum(grid[i][0:curWidth])
            while curSum > k:
                curWidth -= 1
                for j in range(0, i + 1):
                    curSum -= grid[j][curWidth]
            result += curWidth
        return result


# runtime - 19ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        
        sm = 0
        m = len(grid)
        n = len(grid[0])

        ans = 0
        c = n
        for r in range(m):
            sm += sum(grid[r][:c])
            while sm > k:
                c -= 1
                if c < 0:
                    return ans
                for rr in range(r+1):
                    sm -= grid[rr][c]
            ans += c
        return ans


# runtime - 24ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        if not len(grid) : return 0
        r,c=0,len(grid[0])
        ans=s=0
        for r,row in enumerate(grid):
            s+= sum(row[:c])
            while s>k and c>0:
                s-=sum ( grid[_][c-1] for _ in range(r+1))
                c-=1
            if s<=k:
                ans+=c
            else:
                break    
        return ans        



# runtime - 30ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        if not len(grid) : return 0
        r,c=0,len(grid[0])
        ans=s=0
        for r,row in enumerate(grid):
            s+= sum(row[:c])
            while s>k and c>0:
                s-=sum ( grid[_][c-1] for _ in range(r+1))
                c-=1
            if s>k:
                break
            ans+=c        
        return ans        


# runtime - 35ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        s = 0
        ans = 0
        n = len(grid)
        m = len(grid[0])
        right = m-1
        for i in range(n):
            for j in range(right+1):
                s += grid[i][j]
            while right >= 0 and s > k:
                for ii in range(i+1):
                    s -= grid[ii][right]
                right -= 1
            ans += right + 1
        return ans


# runtime - 41ms
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        s = 0
        ans = 0
        n = len(grid)
        m = len(grid[0])
        right = m-1
        for i in range(n):
            for j in range(right+1):
                s += grid[i][j]
            while right >= 0 and s > k:
                for ii in range(i+1):
                    s -= grid[ii][right]
                right -= 1
            ans += right + 1
        return ans
